lib/runtime/src/health_check.rs

```diff
@@ -0,0 +1,523 @@
+    pub request_timeout: Duration,
+}
+
+impl Default for HealthCheckConfig {
+    fn default() -> Self {
+        Self {
+            canary_wait_time: Duration::from_secs(crate::config::DEFAULT_CANARY_WAIT_TIME_SECS),
+            request_timeout: Duration::from_secs(
+                crate::config::DEFAULT_HEALTH_CHECK_REQUEST_TIMEOUT_SECS,
+            ),
+        }
+    }
+}
+
+// Type alias for the router cache to improve readability
+// Maps endpoint subject -> router and payload
+type RouterCache =
+    Arc<Mutex<HashMap<String, Arc<PushRouter<serde_json::Value, Annotated<serde_json::Value>>>>>>;
+
+/// Health check manager that monitors endpoint health
+pub struct HealthCheckManager {
+    drt: DistributedRuntime,
+    config: HealthCheckConfig,
+    /// Cache of PushRouters and payloads for each endpoint
+    router_cache: RouterCache,
+    /// Track per-endpoint health check tasks  
+    /// Maps: endpoint_subject -> task_handle
+    endpoint_tasks: Arc<Mutex<HashMap<String, JoinHandle<()>>>>,
+}
+
+impl HealthCheckManager {
+    pub fn new(drt: DistributedRuntime, config: HealthCheckConfig) -> Self {
+        Self {
+            drt,
+            config,
+            router_cache: Arc::new(Mutex::new(HashMap::new())),
+            endpoint_tasks: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    /// Get or create a PushRouter for an endpoint
+    async fn get_or_create_router(
+        &self,
+        endpoint: Endpoint,
+    ) -> anyhow::Result<Arc<PushRouter<serde_json::Value, Annotated<serde_json::Value>>>> {
+        // TODO: Maybe use Endpoint::etcd_root()?
+        let cache_key = endpoint.path(); // Use the full endpoint path as cache key
+
+        // Check cache first
+        {
+            let cache = self.router_cache.lock().unwrap();
+            if let Some(router) = cache.get(&cache_key) {
+                return Ok(router.clone());
+            }
+        }
+
+        // Create a client that discovers instances dynamically for this endpoint
+        let client = Client::new_dynamic(endpoint).await?;
+
+        // Create PushRouter - it will use direct routing when we call direct()
+        let router: Arc<PushRouter<serde_json::Value, Annotated<serde_json::Value>>> = Arc::new(
+            PushRouter::from_client(
+                client,
+                crate::pipeline::RouterMode::RoundRobin, // Default mode, we'll use direct() explicitly
+            )
+            .await?,
+        );
+
+        // Cache it
+        self.router_cache
+            .lock()
+            .unwrap()
+            .insert(cache_key, router.clone());
+
+        Ok(router)
+    }
+
+    /// Start the health check manager by spawning per-endpoint monitoring tasks
+    pub async fn start(self: Arc<Self>) -> anyhow::Result<()> {
+        // Get all registered endpoints at startup
+        let targets = {
+            let system_health = self.drt.system_health.lock().unwrap();
+            system_health.get_health_check_targets()
```

Comment:
as a note: we should combine this with the health check ready item - so that we use the same mechanism to enable both - i.e. if you set a health check payload then we use that for the ready as well .... instead of the current environment variable